mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp

```diff
@@ -2848,6 +2848,37 @@ struct LLVMOpAsmDialectInterface : public OpAsmDialectInterface {
 // DialectInlinerInterface
 //===----------------------------------------------------------------------===//
 
+/// Move all `LLVM::AllocaOp` operations with a constant size in `sourceBlock`
+/// to the end of `entryBlock`. If necessary, the operation defining the
+/// constant will be moved as well.
+static void moveConstantAllocasToEntryBlock(Block *sourceBlock,
+                                            Block *entryBlock) {
+  assert(entryBlock->isEntryBlock() &&
+         "Second argument must be an entry block.");
+  assert(sourceBlock->getParent() == entryBlock->getParent() &&
+         "Blocks must belong to the same region.");
+  if (sourceBlock == entryBlock)
+    // Nothing to do.
+    return;
+  SmallVector<std::pair<LLVM::AllocaOp, IntegerAttr>> allocasToMove;
+  // Don't walk nested regions of the block, since we can't be sure of their
+  // semantics.
+  for (auto allocaOp : sourceBlock->getOps<LLVM::AllocaOp>()) {
+    IntegerAttr arraySize;
+    if (matchPattern(allocaOp.getArraySize(), m_Constant(&arraySize)))
+      allocasToMove.emplace_back(allocaOp, arraySize);
+  }
+  // Cannot use `back()`, since the constructor of `OpBuilder` needs an
+  // iterator.
+  OpBuilder builder(entryBlock, --entryBlock->getOperations().end());
```

Comment:
Would it make sense to insert from the beginning (entryBlock->begin())?

Reply:
We can also do that, it's pretty arbitrary I guess. Intuitively I would expect stuff to be appended to the end, but if you think it's better I'll put it to the beginning?

Reply:
I am just worried that the -- may not work if the entry block is empty? Not sure though.